# Segmenting and Clustering Toronto Neighborhoods

### Let's Import the Libraries we will need

In [3]:
# The code was removed by Watson Studio for sharing.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    ca-certificates-2019.11.28 |       hecc5488_0         145 KB  conda-forge
    certifi-2019.11.28         |           py36_0         149 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1.20.0-py_0       conda-forge

The following packages will be UPDATED:

    ca-

## Next, we read the csv file that contains the Toronto neighborhoods with geospatial data into a Pandas Dataframe.

In [4]:
df1= pd.read_csv(project.get_file('toronto_geo.csv'))

df1.head()

Postcode      Borough                          Neighborhood   Latitude  \
0      M1B  Scarborough                         Rouge,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2      M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3      M1G  Scarborough                                Woburn  43.770992   
4      M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

### Let's examine the data by seeing how many boroughs and neigborhoods we have.

In [11]:
print('We have {} Boroughs and {} Neighborhoods'.format(len(df1['Borough'].unique()),len(df1['Neighborhood'])))

We have 11 Boroughs and 103 Neighborhoods


### Now let's get a better look at Toronto by creating a map.

#### First we need the coordinates of Toronto.

In [9]:
address = 'Toronto, Ontario'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.20.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinates of Toronto are 43.653963, -79.387207.


#### Now we can create our map.

In [10]:
 Toronto_map = folium.Map(location=[latitude, longitude], zoom_start=10)

# Using the geospatial data from our dataframe, we add markers on the map for each neighborhood

for lat, lng, borough, neighborhood in zip(df1['Latitude'], df1['Longitude'], df1['Borough'], df1['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough) # This will label the markers by neighborhood, borough name
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(Toronto_map)  
    
Toronto_map

## Let's analyze the Toronto neighborhoods

##### We will focus on "Downtown Toronto" as the borough of interest and then cluster the neighborhoods around it.

##### First, let's sort the Boroughs that contain the word 'Toronto'.

In [58]:
toronto = df1[df1['Borough'].str.contains('Toronto')].reset_index(drop = True)

toronto.head()

Postcode          Borough                   Neighborhood   Latitude  \
0      M4E     East Toronto                    The Beaches  43.676357   
1      M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2      M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3      M4M     East Toronto                Studio District  43.659526   
4      M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

#### Now let's get the coordinates for Downtown Toronto

In [14]:
address = 'Downtown Toronto, TO'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Downtown Toronto are {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.20.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The coordinates of Downtown Toronto are 43.6541737, -79.3808116451341.


#### Let's create a new map so we can better visualize our data.

In [17]:

downtown_map = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, lng, label in zip(toronto['Latitude'], toronto['Longitude'], toronto['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(downtown_map)  
    
downtown_map

## Now let's explore the Downtown Toronto neighborhoods using the FourSquare API

In [19]:
# The code was removed by Watson Studio for sharing.

##### The FourSquare API uses the coordinates of the neghborhoods instead of the names. However, some of the neighborhoods have multiple names appended together as one name. 

##### Due to this, we will create a function that will do the folowing:

- Extracts name, latitude and longitude of the neighborhoods from our dataframe
- Inputs the client credentials, name, latitude, longitude to the Foursquare API to explore venues within 500m radius of the neighborhood
- Retreives Venue data from the Json code that FourSquare API outputs
- Creates a dataframe of all the relevant information we need to analyze the venues

In [20]:
def NearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [21]:
radius = 500
LIMIT = 100 # this will limit our results to 100 venues per neighborhood
Toronto_venues = NearbyVenues(names = toronto['Neighborhood'], latitudes = toronto['Latitude'], longitudes = toronto['Longitude'])

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvalles
Runnymede

### Now let's examine the dataset from the Toronto venues.

In [22]:
print(Toronto_venues.shape)
Toronto_venues.head(10)

(1685, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                  The Beaches              43.676357              -79.293031   
1                  The Beaches              43.676357              -79.293031   
2                  The Beaches              43.676357              -79.293031   
3                  The Beaches              43.676357              -79.293031   
4  The Danforth West,Riverdale              43.679557              -79.352188   
5  The Danforth West,Riverdale              43.679557              -79.352188   
6  The Danforth West,Riverdale              43.679557              -79.352188   
7  The Danforth West,Riverdale              43.679557              -79.352188   
8  The Danforth West,Riverdale              43.679557              -79.352188   
9  The Danforth West,Riverdale              43.679557              -79.352188   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                            Pantheon       43.677621       -79.351434   
5                        Dolce Gelato       43.677773       -79.351187   
6                       MenEssentials       43.677820       -79.351265   
7                     Cafe Fiorentina       43.677743       -79.350115   
8                               Mezes       43.677962       -79.350196   
9                Moksha Yoga Danforth       43.677622       -79.352116   

       Venue Category  
0               Trail  
1   Health Food Store  
2                 Pub  
3        Neighborhood  
4    Greek Restaurant  
5      Ice Cream Shop  
6      Cosmetics Shop  
7  Italian Restaurant  
8    Greek Restaurant  
9         Yoga Studio

#### We have a total of 1685 venues in our neighborhoods within our dataset. The dataframe generated provides detailed information of the venues such as their names and what category they belong to.

#### We can leverage that information to categorize the neigborhoods using k_Means clustering to get more insightful data so we can get a better picture of our neighborhoods and what they can offer.

#### Let's see number of venues returned per neighborhood

In [23]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            58   
Brockton,Exhibition Place,Parkdale Village                             22   
Business Reply Mail Processing Centre 969 Eastern                      17   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     15   
Cabbagetown,St. James Town                                             46   
Central Bay Street                                                     85   
Chinatown,Grange Park,Kensington Market                                91   
Christie                                                               17   
Church and Wellesley                                                   86   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             32   
Davisville North                                                        8   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     15   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            19   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      5   
Harbord,University of Toronto                                          34   
Harbourfront                                                           48   
Harbourfront East,Toronto Islands,Union Station                       100   
High Park,The Junction South                                           24   
Lawrence Park                                                           4   
Little Portugal,Trinity                                                54   
Moore Park,Summerhill East                                              3   
North Toronto West                                                     21   
Parkdale,Roncesvalles                                                  13   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede,Swansea                                                      34   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        98   
Studio District                                                        42   
The Annex,North Midtown,Yorkville                                      23   
The Beaches                                                             4   
The Beaches West,India Bazaar                                          19   
The Danforth West,Riverdale                                            42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             58   
Brockton,Exhibition Place,Parkdale Village                              22   
Business Reply Mail Processing Centre 969 Eastern                       17   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      15   
Cabbagetown,St. James Town                                              46   
Central Bay Street                                                      85   
Chinatown,Grange Park,Kensington Market                                 91   
Christie                                                                17   
Church and Wellesley                                                    86   
Commerce Court,Victoria Hotel      

#### We might be able to use the venue categories and how frequent they appear in the neighborhoods as a way to cluster our data.

##### Let's now see how many unique venue categories there are and how frequent their occurence is within the neighborhoods.

In [25]:
Toronto_venues.groupby('Venue Category').count()

Neighborhood  Neighborhood Latitude  \
Venue Category                                                         
Afghan Restaurant                           1                      1   
Airport                                     1                      1   
Airport Food Court                          1                      1   
Airport Lounge                              2                      2   
Airport Service                             3                      3   
Airport Terminal                            2                      2   
American Restaurant                        21                     21   
Antique Shop                                2                      2   
Aquarium                                    5                      5   
Art Gallery                                12                     12   
Arts & Crafts Store                         5                      5   
Asian Restaurant                           14                     14   
Athletics & Sports                          1                      1   
Auto Workshop                               1                      1   
BBQ Joint                                   5                      5   
Baby Store                                  1                      1   
Bagel Shop                                  4                      4   
Bakery                                     42                     42   
Bank                                        5                      5   
Bar                                        37                     37   
Baseball Stadium                            2                      2   
Basketball Stadium                          4                      4   
Beach                                       1                      1   
Bed & Breakfast                             1                      1   
Beer Bar                                   18                     18   
Beer Store                                  4                      4   
Belgian Restaurant                          1                      1   
Bike Rental / Bike Share                    1                      1   
Bistro                                      3                      3   
Boat or Ferry                               1                      1   
Bookstore                                  18                     18   
Boutique                                    2                      2   
Brazilian Restaurant                        3                      3   
Breakfast Spot                             18                     18   
Brewery                                    12                     12   
Bubble Tea Shop                             9                      9   
Building                                    3                      3   
Burger Joint                               21                     21   
Burrito Place                               8                      8   
Bus Line                                    2                      2   
Butcher                                     4                      4   
Café                                       90                     90   
Cajun / Creole Restaurant                   1                      1   
Camera Store                                1                      1   
Candy Store                                 1                      1   
Caribbean Restaurant                        5                      5   
Cheese Shop                                 7                      7   
Chinese Restaurant                         12                     12   
Chocolate Shop                              1                      1   
Church                                      2                      2   
Climbing Gym                                1                      1   
Clothing Store                             18                     18   
Cocktail Bar                               13                     13   
Coffee Shop                               144                    144   
C

##### We can see from the view above that venue category frequency appears to be the best way to define the category of the neighborhood as far as clustering is concerned because the more frequent the category appears in a particular neighborhood, the more likely that neighborhood would be best suited to provide those particular services or experiences to anyone that would want to visit the area. 

##### So we will cluster the neighborhoods on the basis of the frequency of the venue categories.

##### Lets find out how many unique categories we have.

In [26]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 236 uniques categories.


### We now need to find the frequency of occurence of each category in a neighborhood

#### For that, we first need to convert the venue categories to boolean values using one-hot encoding so that if a venue category occurs in a particular neighborhood, we assign 1 to that cell as true, and 0 for the cells where it doesn't occur as false.

In [33]:
# one hot encoding
one_hot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
one_hot['Neighborhoods'] = Toronto_venues['Neighborhood'] 

# I labeled the new column neighborhoods as there is already a column named 'Neighborhood' in the venue categories.

# move neighborhood column to the first column
fixed_columns = [one_hot.columns[-1]] + list(one_hot.columns[:-1])
one_hot = one_hot[fixed_columns] 

print('the shape of one_hot is: ',one_hot.shape)

one_hot.head()

the shape of one_hot is:  (1685, 237)


Neighborhoods  Afghan Restaurant  Airport  \
0                  The Beaches                  0        0   
1                  The Beaches                  0        0   
2                  The Beaches                  0        0   
3                  The Beaches                  0        0   
4  The Danforth West,Riverdale                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                    0             0         0            0   
1                    0             0         0            0   
2                    0             0         0            0   
3                    0             0         0            0   
4                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Rental / Bike Share  Bistro  Boat or Ferry  \
0                   0                         0       0              0   
1                   0                         0       0              0   
2                   0                         0       0              0   
3                   0                         0       0              0   
4                   0                         0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  Butcher  \
0                0         0             0              0         0        0   
1                0         0             0              0         0        0   
2                0         0             0              0         0        0   
3                0         0             0              0         0        0   
4                0         0             0              0         0        0   

   Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
0     0      

###  Find the mean frequency of occurence of each category in a neighborhood so we can use K-Clustering

In [37]:
toronto_group = one_hot.groupby('Neighborhoods').mean().reset_index()

toronto_group.head()

Neighborhoods  Afghan Restaurant  \
0                             Adelaide,King,Richmond                0.0   
1                                        Berczy Park                0.0   
2         Brockton,Exhibition Place,Parkdale Village                0.0   
3  Business Reply Mail Processing Centre 969 Eastern                0.0   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...                0.0   

    Airport  Airport Food Court  Airport Lounge  Airport Service  \
0  0.000000            0.000000        0.000000              0.0   
1  0.000000            0.000000        0.000000              0.0   
2  0.000000            0.000000        0.000000              0.0   
3  0.000000            0.000000        0.000000              0.0   
4  0.066667            0.066667        0.133333              0.2   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0          0.000000                 0.02           0.0       0.0     0.010000   
1          0.000000                 0.00           0.0       0.0     0.017241   
2          0.000000                 0.00           0.0       0.0     0.000000   
3          0.000000                 0.00           0.0       0.0     0.000000   
4          0.133333                 0.00           0.0       0.0     0.000000   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                  0.0              0.03                 0.0       0.000000   
1                  0.0              0.00                 0.0       0.000000   
2                  0.0              0.00                 0.0       0.000000   
3                  0.0              0.00                 0.0       0.058824   
4                  0.0              0.00                 0.0       0.000000   

   BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank       Bar  \
0   0.000000         0.0    0.000000  0.030000   0.0  0.040000   
1   0.017241         0.0    0.017241  0.034483   0.0  0.000000   
2   0.000000         0.0    0.000000  0.045455   0.0  0.045455   
3   0.000000         0.0    0.000000  0.000000   0.0  0.000000   
4   0.000000         0.0    0.000000  0.000000   0.0  0.066667   

   Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  Beer Bar  \
0               0.0            0.000000  0.000000              0.0  0.000000   
1               0.0            0.017241  0.017241              0.0  0.034483   
2               0.0            0.000000  0.000000              0.0  0.000000   
3               0.0            0.000000  0.000000              0.0  0.000000   
4               0.0            0.000000  0.000000              0.0  0.000000   

   Beer Store  Belgian Restaurant  Bike Rental / Bike Share    Bistro  \
0         0.0                 0.0                       0.0  0.000000   
1         0.0                 0.0                       0.0  0.017241   
2         0.0                 0.0                       0.0  0.000000   
3         0.0                 0.0                       0.0  0.000000   
4         0.0                 0.0                       0.0  0.000000   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0       0.000000       0.01  0.000000                  0.01        0.010000   
1       0.000000       0.00  0.000000                  0.00        0.017241   
2       0.000000       0.00  0.000000                  0.00        0.090909   
3       0.000000       0.00  0.000000                  0.00        0.000000   
4       0.066667       0.00  0.066667                  0.00        0.000000   

    Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0  0.000000              0.0      0.01          0.03       0.010000       0.0   
1  0.000000              0.0      0.00          0.00       0.000000       0.0   
2  0.000000              0.0      0.00          0.00       0.045455       0.0   
3  0.058824              0.0      0.00          0.00       0.058824       0.0   
4  0.000000         

#### From the shape of toronto_group we can see that we have a total of 38 unique neighborhoods with 237 venue categories

In [38]:
print('the shape of toronto_group is: ', torronto_group.shape)

the shape of toronto_group is:  (38, 237)


### Now let's print each neighborhood with the top 5 most frequent venues.

In [39]:
top_venues = 5

for hood in toronto_group['Neighborhoods']:
    print("----"+hood+"----")
    temp = toronto_group[toronto_group['Neighborhoods'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.08
1             Café  0.05
2       Steakhouse  0.04
3              Bar  0.04
4  Thai Restaurant  0.03


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2            Beer Bar  0.03
3  Seafood Restaurant  0.03
4   French Restaurant  0.03


----Brockton,Exhibition Place,Parkdale Village----
            venue  freq
0            Café  0.14
1  Breakfast Spot  0.09
2     Coffee Shop  0.09
3     Yoga Studio  0.05
4      Restaurant  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.12
1         Yoga Studio  0.06
2       Garden Center  0.06
3         Pizza Place  0.06
4          Comic Shop  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0   Airport Service  0.20
1    Airport Lounge  0.13
2  Airport Term

### Let's put that into a Pandas DataFrame

##### First, let's write a function to sort the venues in descending order.

In [40]:
def return_most_common_venues(row, top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:top_venues]

##### Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [41]:
top_venues = 10

indicators = ['st', 'nd', 'rd']

# Label columns as 1st, 2nd and so on
columns = ['Neighborhoods']
for ind in np.arange(top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind])) # for 1st, 2nd, 3rd
    except:
        columns.append('{}th Most Common Venue'.format(ind+1)) ### for 4th, 5th,...

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns) ## assign column names we just created to a new dataframe
neighborhoods_venues_sorted['Neighborhoods'] = toronto_group['Neighborhoods']## add neighborhoods column

for ind in np.arange(toronto_group.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_group.iloc[ind, :], top_venues)

neighborhoods_venues_sorted.head()

Neighborhoods 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                   Bar            Steakhouse   
1          Cocktail Bar                  Café                Bakery   
2           Coffee Shop        Breakfast Spot           Yoga Studio   
3           Yoga Studio         Auto Workshop           Pizza Place   
4        Airport Lounge      Airport Terminal       Harbor / Marina   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Burger Joint            Restaurant                Bakery   
1           Cheese Shop    Seafood Restaurant              Beer Bar   
2               Stadium         Burrito Place            Restaurant   
3      Recording Studio            Restaurant               Butcher   
4                   Bar                 Plane      Sculpture Garden   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Sushi Restaurant      Asian Restaurant        Thai Restaurant  
1     French Restaurant        Farmers Market             Steakhouse  
2          Climbing Gym             Pet Store                 Bakery  
3         Burrito Place               Brewery             Skate Park  
4         Boat or Ferry              Boutique                Airport

## Time to cluster the neighborhoods

In [42]:
# set number of clusters
k = 5

toronto_group_clustered = toronto_group.drop('Neighborhoods', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k, random_state=0).fit(toronto_group_clustered)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:38] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0,
       3, 0, 1, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 4, 0, 0], dtype=int32)

##### Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [66]:
toronto_merged = toronto

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhoods'), on='Neighborhood')

toronto_merged # check the last columns!

Postcode           Borough  \
0       M4E      East Toronto   
1       M4K      East Toronto   
2       M4L      East Toronto   
3       M4M      East Toronto   
4       M4N   Central Toronto   
5       M4P   Central Toronto   
6       M4R   Central Toronto   
7       M4S   Central Toronto   
8       M4T   Central Toronto   
9       M4V   Central Toronto   
10      M4W  Downtown Toronto   
11      M4X  Downtown Toronto   
12      M4Y  Downtown Toronto   
13      M5A  Downtown Toronto   
14      M5B  Downtown Toronto   
15      M5C  Downtown Toronto   
16      M5E  Downtown Toronto   
17      M5G  Downtown Toronto   
18      M5H  Downtown Toronto   
19      M5J  Downtown Toronto   
20      M5K  Downtown Toronto   
21      M5L  Downtown Toronto   
22      M5N   Central Toronto   
23      M5P   Central Toronto   
24      M5R   Central Toronto   
25      M5S  Downtown Toronto   
26      M5T  Downtown Toronto   
27      M5V  Downtown Toronto   
28      M5W  Downtown Toronto   
29      M5X  Downtown Toronto   
30      M6G  Downtown Toronto   
31      M6H      West Toronto   
32      M6J      West Toronto   
33      M6K      West Toronto   
34      M6P      West Toronto   
35      M6R      West Toronto   
36      M6S      West Toronto   
37      M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                                         The Beaches  43.676357 -79.293031   
1                         The Danforth West,Riverdale  43.679557 -79.352188   
2                       The Beaches West,India Bazaar  43.668999 -79.315572   
3                                     Studio District  43.659526 -79.340923   
4                                       Lawrence Park  43.728020 -79.388790   
5                                    Davisville North  43.712751 -79.390197   
6                                  North Toronto West  43.715383 -79.405678   
7                                          Davisville  43.704324 -79.388790   
8                          Moore Park,Summerhill East  43.689574 -79.383160   
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049   
10                                           Rosedale  43.679563 -79.377529   
11                         Cabbagetown,St. James Town  43.667967 -79.367675   
12                               Church and Wellesley  43.665860 -79.383160   
13                                       Harbourfront  43.654260 -79.360636   
14                            Ryerson,Garden District  43.657162 -79.378937   
15                                     St. James Town  43.651494 -79.375418   
16                                        Berczy Park  43.644771 -79.373306   
17                                 Central Bay Street  43.657952 -79.387383   
18                             Adelaide,King,Richmond  43.650571 -79.384568   
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752   
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576   
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817   
22                                           Roselawn  43.711695 -79.416936   
23                 Forest Hill North,Forest Hill West  43.696948 -79.411307   
24                  The Annex,North Midtown,Yorkville  43.672710 -79.405678   
25                      Harbord,University of Toronto  43.662696 -79.400049   
26            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049   
27  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
29              First Canadian Place,Underground city  43.648429 -79.382280   
30                                           Christie  43.669542 -79.422564   
31                        Dovercourt Village,Dufferin  43.669005 -79.442259   
32                            Little Portugal,Trinity  43.647927 -79.419750   
33         Brockton,Exhibitio

## Visualizing the clusters on a map

### Let's visualize the clusters using Folium library.

In [70]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k)
ys = [i+x+(i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Now we can analyze the clusters

### Let's start with the cluster with the highest frequency

In [71]:
# Cluster 0 = Red, since red has the highes frequency, we will start there.

toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]] #red

Borough  Cluster Labels  1st Most Common Venue  \
0       East Toronto               0      Health Food Store   
1       East Toronto               0       Greek Restaurant   
2       East Toronto               0         Sandwich Place   
3       East Toronto               0                   Café   
4    Central Toronto               0                   Park   
5    Central Toronto               0         Sandwich Place   
6    Central Toronto               0            Coffee Shop   
7    Central Toronto               0           Dessert Shop   
8    Central Toronto               0             Playground   
9    Central Toronto               0                    Pub   
10  Downtown Toronto               0                   Park   
11  Downtown Toronto               0            Coffee Shop   
12  Downtown Toronto               0            Coffee Shop   
13  Downtown Toronto               0            Coffee Shop   
14  Downtown Toronto               0            Coffee Shop   
15  Downtown Toronto               0            Coffee Shop   
16  Downtown Toronto               0            Coffee Shop   
18  Downtown Toronto               0            Coffee Shop   
19  Downtown Toronto               0            Coffee Shop   
20  Downtown Toronto               0            Coffee Shop   
21  Downtown Toronto               0            Coffee Shop   
23   Central Toronto               0                  Trail   
25  Downtown Toronto               0                   Café   
26  Downtown Toronto               0  Vietnamese Restaurant   
29  Downtown Toronto               0            Coffee Shop   
30  Downtown Toronto               0          Grocery Store   
31      West Toronto               0                 Bakery   
32      West Toronto               0                    Bar   
33      West Toronto               0                   Café   
34      West Toronto               0     Mexican Restaurant   
36      West Toronto               0                   Café   
37      East Toronto               0     Light Rail Station   

            2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                    Neighborhood                   Pub                 Trail   
1                     Coffee Shop        Ice Cream Shop    Italian Restaurant   
2              Light Rail Station         Movie Theater          Burger Joint   
3                     Coffee Shop                Bakery    Italian Restaurant   
4                            Lake           Swim School              Bus Line   
5                             Gym     Convenience Store        Breakfast Spot   
6             Sporting Goods Shop          Burger Joint    Salon / Barbershop   
7                  Sandwich Place    Italian Restaurant                  Café   
8                            Park          Tennis Court                 Diner   
9                     Coffee Shop            Bagel Shop            Sports Bar   
10                     Playground                 Trail                 Diner   
11                           Café            Restaurant                   Pub   
12            Japanese Restaurant               Gay Bar      Sushi Restaurant   
13                           Park                Bakery                   Pub   
14                 Clothing Store        Cosmetics Shop                  Café   
15                           Café            Restaurant                 Hotel   
16                   Cocktail Bar                  Café                Bakery   
18                           Café                   Bar            Steakhouse   
19                       Aquarium                  Café                 Hotel   
20                           Café                 Hotel            Restaurant   
21                           Café                 Hotel            Restaurant   
23                       Bus Line                  Park      Sushi Restaurant   
25                         Bakery             Bookstore            Res

##### We can see that most top venues for cluster label 0 are either restaurants or cafes which are colored red. Red clusters seem to dominate the map so it appears that Downtown Toronto is a great place to grab a bite to eat wtih a vast variety of multicultural restaurants to choose from. Or get a tasty drink that will keep you going all day. 

### Let's see what the other clusters have to offer

In [72]:
# Cluster 1
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]] # purple

Borough  Cluster Labels 1st Most Common Venue  \
24   Central Toronto               1        Sandwich Place   
27  Downtown Toronto               1       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
24                  Café           Coffee Shop           Pizza Place   
27        Airport Lounge      Airport Terminal       Harbor / Marina   

   5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
24           Shoe Repair          Burger Joint  Middle Eastern Restaurant   
27                   Bar                 Plane           Sculpture Garden   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
24     Indian Restaurant                   Pub              BBQ Joint  
27         Boat or Ferry              Boutique                Airport

##### We can see with the purple cluster that while there are still a great number of restaurants and coffee shops, there are an almost equal amount of venues associate with travel. We have airport services, lounges and terminals alont with Harbors to catch a boat or ferry. 

##### It's widely known that people tend to eat while traveling so that is why we see places to grab a sandwich, pizza or burger that is cose to their chosen mode of transportation.

In [73]:
#Cluster 2

toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]] #blue

Borough  Cluster Labels 1st Most Common Venue  \
28  Downtown Toronto               2           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28                  Café            Restaurant          Cocktail Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28   Japanese Restaurant    Italian Restaurant    Seafood Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28                 Hotel              Beer Bar                   Park

##### What we can see with the blue cluster is that while 'Hotel' appears to be the 8th most common venue, most people that stay at hotels tend to spend most of their time visiting a coffee shop in the morning as the hotel coffee may not be the best and at high-end restaurants and bars at night. 

In [74]:
# Cluster 3
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]] #green

Borough  Cluster Labels 1st Most Common Venue  \
17  Downtown Toronto               3           Coffee Shop   
22   Central Toronto               3                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17    Italian Restaurant                  Café   Japanese Restaurant   
22                  Pool                 Diner        Farmers Market   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17        Sandwich Place          Burger Joint        Ice Cream Shop   
22    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
17                   Bar    Chinese Restaurant                    Juice Bar  
22   Empanada Restaurant     Electronics Store  Eastern European Restaurant

##### What we can gather from the green clusters is that there are a lot of places to go for entertainment and shopping such as swimming, going to the farmer's market or purchase some electronics. And of course there are more restaurants, coffee shops and bars as who doesn't like to engage in those activities while shopping.

In [75]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]] #orange

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
35  West Toronto               4             Gift Shop             Bookstore   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
35          Dessert Shop  Eastern European Restaurant         Movie Theater   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
35                   Bar    Italian Restaurant               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
35            Restaurant         Breakfast Spot

##### It looks like we can infer that the orange cluster is geared more towards shopping and entertainment with a gift shop, bookstore and movie theater within the top 5 venues.